internal/pkg/agent/control/client/client.go

```diff
@@ -250,75 +282,66 @@ func (c *client) Upgrade(ctx context.Context, version string, sourceURI string)
 
-	for _, proc := range resp.Procs {
-		meta := ProcMeta{
-			Process:            proc.Process,
-			Name:               proc.Name,
-			Hostname:           proc.Hostname,
-			ID:                 proc.Id,
-			EphemeralID:        proc.EphemeralId,
-			Version:            proc.Version,
-			BuildCommit:        proc.BuildCommit,
-			Username:           proc.Username,
-			UserID:             proc.UserId,
-			UserGID:            proc.UserGid,
-			BinaryArchitecture: proc.Architecture,
-			RouteKey:           proc.RouteKey,
-			ElasticLicensed:    proc.ElasticLicensed,
-			Error:              proc.Error,
-		}
-		if proc.BuildTime != "" {
-			ts, err := time.Parse(time.RFC3339, proc.BuildTime)
-			if err != nil {
-				if meta.Error != "" {
-					meta.Error += ", " + err.Error()
-				} else {
-					meta.Error = err.Error()
-				}
-			} else {
-				meta.BuildTime = ts
-			}
-		}
-		procMeta = append(procMeta, meta)
+	files := make([]DiagnosticFileResult, 0, len(resp.Results))
+	for _, f := range resp.Results {
+		files = append(files, DiagnosticFileResult{
+			Name:        f.Name,
+			Filename:    f.Filename,
+			Description: f.Description,
+			ContentType: f.ContentType,
+			Content:     f.Content,
+			Generated:   f.Generated.AsTime(),
+		})
 	}
-	return procMeta, nil
+	return files, nil
 }
 
-// Pprof gathers /debug/pprof data and returns a map of pprof-type: ProcPProf data
-func (c *client) Pprof(ctx context.Context, d time.Duration, pprofTypes []cproto.PprofOption, appName, routeKey string) (map[string][]ProcPProf, error) {
-	resp, err := c.client.Pprof(ctx, &cproto.PprofRequest{
-		PprofType:     pprofTypes,
-		TraceDuration: d.String(),
-		AppName:       appName,
-		RouteKey:      routeKey,
-	})
+// DiagnosticUnits gathers diagnostics information from specific units (or all if non are provided).
+func (c *client) DiagnosticUnits(ctx context.Context, units ...DiagnosticUnitRequest) ([]DiagnosticUnitResult, error) {
+	reqs := make([]*cproto.DiagnosticUnitRequest, 0, len(units))
+	for _, u := range units {
+		reqs = append(reqs, &cproto.DiagnosticUnitRequest{
+			UnitType: u.UnitType,
+			UnitId:   u.UnitID,
+		})
+	}
+
+	resp, err := c.client.DiagnosticUnits(ctx, &cproto.DiagnosticUnitsRequest{Units: reqs})
 	if err != nil {
 		return nil, err
 	}
-	res := map[string][]ProcPProf{}
-	for _, pType := range pprofTypes {
-		res[pType.String()] = make([]ProcPProf, 0)
-	}
-	for _, r := range resp.Results {
-		res[r.PprofType.String()] = append(res[r.PprofType.String()], ProcPProf{
-			Name:     r.AppName,
-			RouteKey: r.RouteKey,
-			Result:   r.Result,
-			Error:    r.Error,
+
+	results := make([]DiagnosticUnitResult, 0, len(resp.Units))
+	for _, u := range resp.Units {
+		files := make([]DiagnosticFileResult, 0, len(u.Results))
+		for _, f := range u.Results {
+			files = append(files, DiagnosticFileResult{
+				Name:        f.Name,
+				Filename:    f.Filename,
+				Description: f.Description,
+				ContentType: f.ContentType,
+				Content:     f.Content,
```

Comment:
Since this [is the raw file content](https://github.com/elastic/elastic-agent-client/blob/c29f5dc424c21a2b5a6a6ba48b8fe8d20e2eb8cd/elastic-agent-client.proto#L135-L136) is memory usage a concern here if there are a large number of units? Do we know what the typical size of the Beat diagnostics are?

Reply:
Most are not that large, just a few kb at most. I would think this should not be much of an issue. If your performing diagnostics a jump in memory usage for that information shouldn't be too much.